R/check_target_tbl_colnames.R

```diff
@@ -1,104 +1,164 @@
+    check <- setequal(col_names, expected_cols)
+
+    if (!check) {
+      missing_cols <- setdiff(expected_cols, col_names)
+      extra_cols <- setdiff(col_names, expected_cols)
+
+      if (length(missing_cols) > 0L) {
+        details <- c(
+          details,
+          cli::format_inline(
+            "Required column{?s} {.val {missing_cols}} defined in {.file target-data.json} {?is/are} missing."
+          )
+        )
+      }
+
+      if (length(extra_cols) > 0L) {
+        details <- c(
+          details,
+          cli::format_inline(
+            "Extra column{?s} {.val {extra_cols}} detected that {?is/are} not defined in {.file target-data.json}."
+          )
+        )
+      }
+    }
+  } else {
+    # Inference mode: use existing validation logic
+    config_tasks <- read_config(hub_path)
+    task_ids <- hubUtils::get_task_id_names(config_tasks)
+
+    required <- switch(
+      target_type,
+      `time-series` = c(
+        "observation",
+        get_target_task_id(config_tasks)
+      ),
+      `oracle-output` = c(
+        "oracle_value",
+        get_target_task_id(config_tasks),
+        # output_type and output_type_id are only required for "cdf" and "pmf"
+        # output types.
+        if (any(unique(target_tbl[["output_type"]]) %in% c("cdf", "pmf"))) {
+          hubUtils::std_colnames[c("output_type", "output_type_id")]
+        } else {
+          NULL
+        }
       )
     )
-  }
 
-  if (target_type == "time-series") {
-    has_as_of_col <- "as_of" %in% col_names
-    req_col_n <- length(required) + 1L + has_as_of_col
-    check_col_n <- ncol(target_tbl) >= req_col_n
-    if (!check_col_n) {
+    missing_cols <- setdiff(required, col_names)
+    check_missing <- length(missing_cols) == 0L
+    if (!check_missing) {
       details <- c(
         details,
         cli::format_inline(
-          "Fewer columns ({.val {ncol(target_tbl)}}) than the required number of
-          columns ({.val {req_col_n}}) detected."
+          "Required column{?s} {.val {missing_cols}} {?is/are} missing."
         )
       )
     }
-  } else {
-    check_col_n <- TRUE
-  }
 
-  invalid_cols <- switch(target_type,
-    `time-series` = {
-      # Time series is more lenient as it allows additional columns.
-      # So we check against invalid columns.
-      invalid <- c(
-        "value",
-        "output_type",
-        "output_type_id",
-        "oracle_value"
-      )
-      intersect(col_names, invalid)
-    },
-    `oracle-output` = {
-      # Oracle output is stricter as no additional columns are allowed.
-      # Columns must be a subset of valid task IDs and optionally
-      valid <- c(
-        required,
-        hubUtils::std_colnames[c("output_type", "output_type_id")],
-        task_ids
-      ) |>
-        unique()
-      setdiff(col_names, valid)
+    if (target_type == "time-series") {
+      has_as_of_col <- "as_of" %in% col_names
+      req_col_n <- length(required) + 1L + has_as_of_col
```

Comment:
I am not sure why we add `+ 1L + has_as_of_col` here? Does that mean we expect an additional `as_of` column and another additional column?